### PR TITLE
circuit breaker with half open state

### PR DESCRIPTION
**Pull request by jiangyt-git**
### What problem does this PR solve?
        解决server在夯住的情况下，比如可以响应ping请求、但没法响应业务逻辑请求时，单次ping请求成功就可以从熔断恢复，然后需要重新采样才能继续熔断的问题。
Issue Number:

Problem Summary:

### What is changed and the side effects?

Changed:

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: src/brpc/circuit_breaker.cpp [hunk 1]
```diff
@@ -45,6 +46,10 @@ DEFINE_int32(circuit_breaker_max_isolation_duration_ms, 30000,
     "Maximum isolation duration in milliseconds");
 DEFINE_double(circuit_breaker_epsilon_value, 0.02,
     "ema_alpha = 1 - std::pow(epsilon, 1.0 / window_size)");
+DEFINE_int32(circuit_breaker_half_open_window_size, 10,
```
Comment (wwbmmm) [inline src/brpc/circuit_breaker.cpp:49]: 这个功能应该是可选的吧，比如half_open_window_size为0表示关闭该功能。默认为关闭，和之前的行为保持兼容。

File: src/brpc/circuit_breaker.cpp [hunk 2]
```diff
@@ -45,6 +46,10 @@ DEFINE_int32(circuit_breaker_max_isolation_duration_ms, 30000,
     "Maximum isolation duration in milliseconds");
 DEFINE_double(circuit_breaker_epsilon_value, 0.02,
     "ema_alpha = 1 - std::pow(epsilon, 1.0 / window_size)");
+DEFINE_int32(circuit_breaker_half_open_window_size, 0,
+    "Half open window sample size.");
```
Comment (zyearn) [inline src/brpc/circuit_breaker.cpp:50]: Can this be more descriptive?

**Comment by j05070415** [comment]
谢谢，您的邮件我已收到！

File: src/brpc/circuit_breaker.cpp (hunk 2 above)
Comment (jiangyt-git) [reply src/brpc/circuit_breaker.cpp]: Done

**Comment by wwbmmm** [comment]
LGTM

File: src/brpc/circuit_breaker.cpp [hunk 3]
```diff
@@ -188,6 +197,20 @@ bool CircuitBreaker::OnCallEnd(int error_code, int64_t latency) {
...
+        && _half_open.load(butil::memory_order_relaxed)) {
+        if (error_code != 0) {
+            MarkAsBroken();
+            return false;
+        }
+        if (_half_open_success_count.fetch_add(1, butil::memory_order_relaxed)
+                + 1 == FLAGS_circuit_breaker_half_open_window_size) {
+            _half_open.store(false, butil::memory_order_relaxed);
+            _half_open_success_count.store(0, butil::memory_order_relaxed);
+            return true;
```
Comment (zyearn) [inline src/brpc/circuit_breaker.cpp:210]: Should we remove the return here and let this latency passed to _long_window/_short_window.OnCallEnd below instead?

Comment (jiangyt-git) [reply src/brpc/circuit_breaker.cpp]: Done.

**Pull request merged**